core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java

```
@@ -0,0 +1,218 @@
+
+    public MigrationPredictor(
+            KernelServerManager serverManager, MicroServiceManager microServiceManager) {
+        this.serverManager = serverManager;
+        this.microServiceManager = microServiceManager;
+        predictedTime =
+                new TreeSet<AbstractMap.SimpleEntry<InetSocketAddress, Long>>(
+                        new Comparator<AbstractMap.SimpleEntry<InetSocketAddress, Long>>() {
+                            @Override
+                            public int compare(
+                                    AbstractMap.SimpleEntry<InetSocketAddress, Long> o1,
+                                    AbstractMap.SimpleEntry<InetSocketAddress, Long> o2) {
+                                if (o1.getValue() < o2.getValue()) return 1;
+                                else if (o1.getValue() > o2.getValue()) return -1;
+                                return 0;
+                            }
+                        });
+    }
+
+    /**
+     * Predict best kernel server for replica by constructing heap based on time optimization on
+     * available kernel server
+     *
+     * @param replicaID
+     * @param currentKernelServer
+     * @return kernel server address
+     */
+    public InetSocketAddress getBestKernelServer(
+            ReplicaID replicaID, InetSocketAddress currentKernelServer) {
+        // get all kernel server
+        try {
+            List<InetSocketAddress> kernelServers =
+                    GlobalKernelReferences.nodeServer.oms.getServers(null);
+
+            // predict time optimization for replica on each kernel server and push it into sorted
+            // heap
+            for (InetSocketAddress ks : kernelServers) {
+                predictedTime.add(
+                        new AbstractMap.SimpleEntry<InetSocketAddress, Long>(
+                                ks, getOptimizedTime(replicaID, currentKernelServer, ks)));
+            }
+        } catch (RemoteException e) {
```

Don;'t we need to throw these exceptions for the caller to handle, rather than just log them here?

---

Oh, I see, if an exception occurs, you just return the current node as the best node.  Fair enough.  It seems better to catch this exception higher up the call stanc, and explicitly avoid moving the replica in that code instead.